### PR TITLE
git v2 client for inrepoconfig: allow targeted fetches (2nd attempt)

### PR DESCRIPTION
**Pull request by listx**
This is a reintroduction of bc68007b25 (git v2 client for inrepoconfig: allow targeted fetches, 2023-08-21), which was reverted in fb82b69cc3 (Revert "Merge pull request #30453 from listx/minimal-fetch", 2023-08-22).

This version includes a fix for the timing of the fetch. In bc68007b25 we called ensureCommits() with "repoClient", but _before_ it was cloned (created)! This version fixes that (now obvious) bug. Below is the original commit message of bc68007b25, because the semantics there still apply as is.

Allow fetching a small number of commit SHAs directly if RepoOpts specifies FetchCommits.  This avoids the costly RemoteUpdate call which fetches every remote tracking branch (as well as all new ones that show up on the remote).

Make tag fetching optional because they could be unnecessary (for inrepoconfig, we already check out a working tree based on the baseSHA and headSHAs in prowYAMLGetter).

Also, only perform a fetch once. Previously in ensureCommits we did at least 2 separate fetch operations for every presubmit job, because a presubmit has a baseSHA and a headSHA (PR HEAD). Now we combine them into a single fetch command, so that we talk to the server once instead of twice.

/cc @cjwagner @airbornepony @timwangmusic 

/hold

### CODE REVIEW

**Review by airbornepony — APPROVED** [review]

File: prow/git/v2/client_factory.go [hunk 1]
```diff
@@ -339,21 +347,70 @@ func (c *clientFactory) ClientForWithRepoOpts(org, repo string, repoOpts RepoOpt
...
+	// Here we do a targeted fetch, if the repoOpts for the secondary clone
+	// asked for it.
+	//
+	// TODO(listx): Change ensureCommits() so that it takes a "cacher" interface
+	// instead of a generic RepoClient. However, if ShareObjectsWithSourceRepo
+	// was not set, then we cannot run ensureCommits() with the cacher because
+	// the cacher will run on the primary, not secondary (and without --shared,
+	// this will have no effect on the secondary); we have to handle that case
+	// gracefully (perhaps with some logging also).
+	if needTargetedFetch {
```
Comment (timwangmusic) [inline prow/git/v2/client_factory.go:376]: We can combine these 2 if statements.

Comment (listx) [reply prow/git/v2/client_factory.go:376]: That would not be the best thing in this situation because it would obscure the importance of the `needTargetedFetch` variable.

Although, now that I'm looking at this code again, I think this variable is not necessary because we can just reuse repoOpts.FetchCommits. Will update.

Comment (listx) [reply prow/git/v2/client_factory.go]: Done.

File: prow/git/v2/client_factory.go [hunk 2]
```diff
@@ -339,21 +346,69 @@ func (c *clientFactory) ClientForWithRepoOpts(org, repo string, repoOpts RepoOpt
...
+		// We have cloned the repo previously, but will refresh it. By default
+		// we refresh all refs with a call to `git remote update`.
+		//
+		// This is the default behavior if FetchCommits is empty or nil (i.e.,
+		// when we don't define a targeted list of commits to fetch directly).
+		if len(repoOpts.FetchCommits) == 0 {
+			if err := cacheClientCacher.RemoteUpdate(); err != nil {
+				return nil, err
+			}
 		}
```
Comment (cjwagner) [inline prow/git/v2/client_factory.go:358]: We still ignore `NoFetchTags` in this case. We should at least document this in the GoDoc for the field. https://github.com/kubernetes/test-infra/pull/30453#discussion_r1302255762

File: prow/git/v2/client_factory.go [hunk 3]
```diff
@@ -339,21 +346,69 @@ func (c *clientFactory) ClientForWithRepoOpts(org, repo string, repoOpts RepoOpt
...
+		if exists, _ := repoClient.ObjectExists(commitSHA); exists {
+			continue
+		}
+
+		fetchArgs = append(fetchArgs, commitSHA)
+	}
+
+	if err := repoClient.Fetch(fetchArgs...); err != nil {
+		return fmt.Errorf("failed to fetch %s: %v", fetchArgs, err)
+	}
```
Comment (cjwagner) [inline prow/git/v2/client_factory.go:411]: Is this only fetching between the primary and secondary clones? We need to fetch from upstream to the primary clone then fetch to the secondary clone, right?
Or is it the case that this fetch for the secondary clone actually reaches the upstream because the secondary is a `--shared` repo?

File: prow/git/v2/client_factory.go (hunk 2 above)
Comment (listx) [reply prow/git/v2/client_factory.go:358]: The flag is deprecated in https://github.com/kubernetes/test-infra/pull/30428 (https://github.com/kubernetes/test-infra/pull/30428/commits/5c0e7eca5c89f04b9dfdf5b158202bedb77bd0b2).

File: prow/git/v2/client_factory.go (hunk 3 above)
Comment (listx) [reply prow/git/v2/client_factory.go:411]: > Is this only fetching between the primary and secondary clones? We need to fetch from upstream to the primary clone then fetch to the secondary clone, right?

From our internal logs for the 1st attempt, the fetch will happen but with an explicit `<remote>` argument, where `<remote>` looks like "https://...". So even though the repoClient is the secondary clone, it will construct a fetch command that looks like "git fetch https://... ...".

> Or is it the case that this fetch for the secondary clone actually reaches the upstream because the secondary is a --shared repo?

Without the additional `<remote>` argument, it would only try to fetch from the primary, and the fetch will fail even with --shared.

So in summary it looks like the above code should be OK.

Comment (listx) [reply prow/git/v2/client_factory.go:411]: Gonna rebase this on top of https://github.com/kubernetes/test-infra/pull/30488 to test the "fetch, don't clone from scratch" code path.

Comment (listx) [reply prow/git/v2/client_factory.go:411]: Done.

**Review by cjwagner — APPROVED** [review]
LGTM assuming the fixed integration tests pass.

**Comment by listx** [comment]
Let's wait for https://github.com/kubernetes/test-infra/pull/30489 to demonstrate the new integration test case, before trusting the same test case in this PR.
